### PR TITLE
Add Isosurface mode  for volume density

### PR DESCRIPTION
**Pull request by LesikDee**
## Description

These changes add Isosurface mode for volume density viewing and develop volume rendering algorithm.

Changes are:
1. Improved code style and readability in Volume_frag.glsl
2. Added IsoSurface rendering functionality
3. Added ability to change on IsoSurface mode in Volume Render (to change modes enter in the console "set modes.VD.IsoMode true/false"
4. Support es2015 in VolumeMaterial

## Screenshots
image

image

## Type of changes
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read [CONTRIBUTING](https://github.com/epam/miew/blob/master/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/epam/miew/blob/master/CODE_OF_CONDUCT.md) guides.
- [x] I have followed the code style of this project.
- [x] I have run `yarn run ci`: lint and tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation / The changes do not need docs update.

### CODE REVIEW

File: src/gfx/shaders/VolumeMaterial.js [hunk 1]
```diff
@@ -35,43 +35,46 @@ function overrideUniforms(params, defUniforms) {
   return uniforms;
 }
 
-function BackFacePosMaterial(params) {
-  const settings = {
-    uniforms: overrideUniforms(params, {}),
-    vertexShader: `\
+class BackFacePosMaterial extends THREE.ShaderMaterial {
```
Comment (paulsmirnov) [inline src/gfx/shaders/VolumeMaterial.js]: I do agree with SonarCloud and CodeClimate that in addition to wrapping this in ES2015 classes we should get rid of the duplicated code here. Please check the reports and try to generalize the behavior. You could also put shaders in separate glsl files as we wanted to (there's a separate task for that in the whole project but you "touch" this particular code anyway).

File: src/gfx/shaders/VolumeMaterial.js [hunk 2]
```diff
@@ -91,59 +94,74 @@ void main() {
...
+      side: THREE.FrontSide,
+    };
+
+    super(settings);
+    this.updateDefines();
+  }
+
+  updateDefines() {
+    const defines = {};
+    defines.ISO_MODE = globalSettings.now.modes.VD.IsoMode;
```
Comment (paulsmirnov) [inline src/gfx/shaders/VolumeMaterial.js]: This is [marked as a code smell](https://sonarcloud.io/project/issues?id=epam%3Amiew&issues=AWsD4CWFVDQP4Uuf9RPs&open=AWsD4CWFVDQP4Uuf9RPs&pullRequest=162) and it should be easily fixed. Don't define an empty object, create it in the final form.

File: src/settings.js [hunk 3]
```diff
@@ -435,13 +435,15 @@ const defaults = {
...
      * @property {boolean} frame - flag, that turns on box frame painting
+     * @property {boolean} IsoMode - flag, that turns on IsoSurface mode instead of Volume Rendering
      *
      */
     VD: {
       kSigma: 1.0,
       kSigmaMed: 2.0,
       kSigmaMax: 4.0,
       frame: true,
+      IsoMode: false,
```
Comment (paulsmirnov) [inline src/settings.js]: Should we start it from a lowercase letter as the implicit code style suggests?

File: src/Miew.js [hunk 4]
```diff
@@ -3444,6 +3444,12 @@ Miew.prototype._initOnSettingsChanged = function () {
     this._needRender = true;
   });
 
+  on('modes.VD.IsoMode', () => {
+    const mesh = this._getVolumeVisual().getMesh();
```
Comment (paulsmirnov) [inline src/Miew.js]: What if there's no volume in the scene?

File: src/gfx/shaders/VolumeMaterial.js [hunk 5]
```diff
@@ -2,10 +2,10 @@
 /* eslint-disable guard-for-in */
 import * as THREE from 'three';
 import volumeFrag from './Volume_frag.glsl';
+import globalSettings from '../../settings';
```
Comment (paulsmirnov) [inline src/gfx/shaders/VolumeMaterial.js]: No way 😄 We must use `settings` here and change the local variable name instead.

File: src/gfx/shaders/Volume_frag.glsl [hunk 6]
```diff
@@ -20,8 +20,11 @@ uniform sampler2D _WFFRight;
 
 varying vec4 screenSpacePos;
 
-vec4 sample3DTexture(vec3 texCoord)
-{
+#define stepsCount 600
```
Comment (paulsmirnov) [inline src/gfx/shaders/Volume_frag.glsl]: Shouldn't the constants be all uppercase with an underscore?

File: src/gfx/shaders/Volume_frag.glsl [hunk 7]
```diff
@@ -42,85 +45,90 @@ vec4 sample3DTexture(vec3 texCoord)
...
-  vec3 N;
-  N.x = sample3DTextureInclined(iter + dx).r - sample3DTextureInclined(iter - dx).r;
-  N.y = sample3DTextureInclined(iter + dy).r - sample3DTextureInclined(iter - dy).r;
-  N.z = sample3DTextureInclined(iter + dz).r - sample3DTextureInclined(iter - dz).r;
-  N = normalize(N);
-  float dif = max(0.0, dot(N,dir));
+
+  vec3 coordInc = vec3(
+    // #Opt: coordInc.x:(iter + dx).x > 1. ? 0.: sample3DTextureInclined(iter + dx).x,
+    mix(sample3DTextureInclined(iter + dx).x, 0., floor((iter + dx).x)),
```
Comment (paulsmirnov) [inline src/gfx/shaders/Volume_frag.glsl]: Please forgive me this recurring question but isn't it possible to vectorize the calculations? I'm also a little bit concerned with the growth of calculations while we need to reduce/optimize them.

File: src/settings.js (hunk 3 above)
Comment (LesikDee) [reply src/settings.js]: Of course, just a misprint

File: src/Miew.js (hunk 4 above)
Comment (LesikDee) [reply src/Miew.js]: Good comment, thanks, this case should be handled. Resolved now!

File: src/gfx/shaders/VolumeMaterial.js (hunk 5 above)
Comment (LesikDee) [reply src/gfx/shaders/VolumeMaterial.js]: Know that now

File: src/gfx/shaders/Volume_frag.glsl (hunk 7 above)
Comment (LesikDee) [reply src/gfx/shaders/Volume_frag.glsl]: I rewrote this part of the code so now it theoretically should work not worse than before, but in fact, even under heavy load, I couldn't see the difference.

**Pull request merged**